aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu

```diff
@@ -0,0 +1,521 @@
+    const Tensor& gradOutput_,
+    const Tensor& input) {
+  TensorArg grad_input_arg{gradInput, "gradInput", 1};
+  TensorArg grad_output_arg{gradOutput_, "gradOutput_", 2};
+  TensorArg input_arg{input, "input", 3};
+
+  checkAllSameGPU(
+      "adaptive_avg_pool3d_out_cuda",
+      {grad_input_arg, grad_output_arg, input_arg});
+
+  const Tensor gradOutput = gradOutput_.contiguous();
+
+  gradInput.resize_as_(input);
+  gradInput.zero_();
+
+  int64_t sizeD, isizeT, isizeH, isizeW;
+  int64_t osizeT, osizeH, osizeW;
+  int64_t totalZ;
+
+  if (input.ndimension() == 4) {
+    sizeD = input.size(0);
+    isizeT = input.size(1);
+    isizeH = input.size(2);
+    isizeW = input.size(3);
+
+    osizeT = gradOutput.size(1);
+    osizeH = gradOutput.size(2);
+    osizeW = gradOutput.size(3);
+  } else {
+    sizeD = input.size(1);
+    isizeT = input.size(2);
+    isizeH = input.size(3);
+    isizeW = input.size(4);
+
+    osizeT = gradOutput.size(2);
+    osizeH = gradOutput.size(3);
+    osizeW = gradOutput.size(4);
+  }
+
+  // somehow nonatomic is passing all test for volumetric case. (copied from
+  // THCUNN source code.)
+  // bool atomic = (isizeW%osizeW != 0) || (isizeH%osizeH != 0) || // (isizeT%osizeT != 0)
+  bool atomic = false;
+
+  if (input.ndimension() == 4) {
+    totalZ = sizeD * osizeT;
+  } else {
+    int sizeB = input.size(0);
+    totalZ = sizeB * sizeD * osizeT;
+  }
+
+  if (atomic) {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+        input.scalar_type(), "adaptive_avg_pool3d_backward_cuda", [&] {
+          scalar_t* gradInput_data = gradInput.data<scalar_t>();
+          scalar_t* gradOutput_data = gradOutput.data<scalar_t>();
+
+          atomicadaptiveaveragegradinput_loop(
+              gradInput_data, gradOutput_data,
+              totalZ,
+              isizeT, isizeH, isizeW,
+              osizeT, osizeH, osizeW);
+        });
+  } else {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+        input.scalar_type(), "adaptive_avg_pool3d_backward_cuda", [&] {
+          scalar_t* gradInput_data = gradInput.data<scalar_t>();
+          scalar_t* gradOutput_data = gradOutput.data<scalar_t>();
+
+          adaptiveaveragegradinput_loop(
+              gradInput_data, gradOutput_data,
+              totalZ,
+              isizeT, isizeH, isizeW,
+              osizeT, osizeH, osizeW);
+        });
+  }
+}
+
+} // namespace
+
+Tensor& adaptive_avg_pool3d_out_cuda(
+    Tensor& output,
+    const Tensor& input,
+    IntArrayRef output_size) {
+  adaptive_avg_pool3d_out_cuda_template(output, input, output_size);
+  return output;
+}
+
+Tensor adaptive_avg_pool3d_cuda(
+    const Tensor& input,
+    IntArrayRef output_size) {
+  auto output = at::empty({0}, input.options());
+  adaptive_avg_pool3d_out_cuda_template(output, input, output_size);
+  return output;
+}
+
+Tensor& adaptive_avg_pool3d_backward_out_cuda(
+    Tensor& gradInput,
+    const Tensor& gradOutput_,
+    const Tensor& input) {
+  gradInput.resize_as_(input);
```

Comment:
Resizing twice, also in the following function.

Reply:
ah oh, thank you!